### DISCUSSION TITLE
Questions about jacobian debugging.

### DISCUSSION THREAD

**Discussion by Raven-pro** [idaholab/moose#22233] [Q&A General] [answered]
Hello Everyone!
I am trying to solve a two-phase problem based on the three basic equations. As the size of the problem is relatively small(1.5m, 1D channel), a time step needs to  be small(like 1e-5 s). But I need to run the simulation to reach about 30s, so the number of time steps is so large. 
Then here comes the problem. The simulation could converge successfully in about the first 200 time steps, and then it didn't converge. So when it comes to the convergence problems, the main point always lies in Jacobian Matrix, so I used the jacobiandebug tool to analyze the codes. 
And the result were like that. 
```
<frozen importlib._bootstrap>:241: RuntimeWarning: compiletime version 3.9 of module 'hit' does not match runtime version 3.1
Running input with executable ../hmm-opt ...

No errors detected. :-)

Kernel for variable 'W':
  (1,2) Off-diagonal Jacobian for variable 'H' needs to be implemented

Kernel for variable 'H':
  (2,2) On-diagonal Jacobian is slightly off (by 0.091803 %)

Kernel for variable 'P':
  (0,2) Off-diagonal Jacobian for variable 'H' needs to be implemented

Kernel for variable 'W':
  (1,2) Off-diagonal Jacobian for variable 'H' needs to be implemented

Kernel for variable 'H':
  (2,2) On-diagonal Jacobian is slightly off (by 0.158564 %)

Kernel for variable 'P':
  (0,2) Off-diagonal Jacobian for variable 'H' needs to be implemented

Kernel for variable 'W':
  (1,2) Off-diagonal Jacobian for variable 'H' needs to be implemented

Kernel for variable 'H':
  (2,2) On-diagonal Jacobian is slightly off (by 0.206703 %)

Kernel for variable 'P':
  (0,2) Off-diagonal Jacobian for variable 'H' needs to be implemented

Kernel for variable 'W':
  (1,2) Off-diagonal Jacobian for variable 'H' needs to be implemented

Kernel for variable 'H':
  (2,2) On-diagonal Jacobian is slightly off (by 0.240198 %)

Kernel for variable 'P':
  (0,2) Off-diagonal Jacobian for variable 'H' needs to be implemented

Kernel for variable 'W':
  (1,2) Off-diagonal Jacobian for variable 'H' needs to be implemented

Kernel for variable 'H':
  (2,2) On-diagonal Jacobian is slightly off (by 0.262970 %)

Kernel for variable 'P':
  (0,2) Off-diagonal Jacobian for variable 'H' needs to be implemented

Kernel for variable 'W':
  (1,2) Off-diagonal Jacobian for variable 'H' needs to be implemented

Kernel for variable 'H':
  (2,2) On-diagonal Jacobian is slightly off (by 0.276401 %)

Kernel for variable 'P':
  (0,2) Off-diagonal Jacobian for variable 'H' needs to be implemented

Kernel for variable 'W':
  (1,2) Off-diagonal Jacobian for variable 'H' needs to be implemented

Kernel for variable 'H':
  (2,2) On-diagonal Jacobian is slightly off (by 0.283936 %)

Kernel for variable 'P':
  (0,2) Off-diagonal Jacobian for variable 'H' needs to be implemented

Kernel for variable 'W':
  (1,2) Off-diagonal Jacobian for variable 'H' needs to be implemented

Kernel for variable 'H':
  (2,2) On-diagonal Jacobian is slightly off (by 0.286852 %)

Kernel for variable 'P':
  (0,2) Off-diagonal Jacobian for variable 'H' needs to be implemented

Kernel for variable 'W':
  (1,2) Off-diagonal Jacobian for variable 'H' needs to be implemented

Kernel for variable 'H':
  (2,2) On-diagonal Jacobian is slightly off (by 0.286377 %)

Kernel for variable 'P':
  (0,2) Off-diagonal Jacobian for variable 'H' needs to be implemented

Kernel for variable 'W':
  (1,2) Off-diagonal Jacobian for variable 'H' needs to be implemented

Kernel for variable 'H':
  (2,2) On-diagonal Jacobian is slightly off (by 0.283390 %)

Kernel for variable 'P':
  (0,2) Off-diagonal Jacobian for variable 'H' needs to be implemented
```
These are just a little part of messages output, and the rest are some how the same as what I am showing here.
So I see there might be some problems in Jacobian, leading to not converging of my code, but I don't know the meaning of the messages and what I can do to fix the problem, can anybody tell me about how to debug based on the messages of jacobiananalizer? 
And I also want to know if there is any method to output the jacobian and condition number so I can check if my jacobian is right.
Thank you so much!

**Comment by GiudGiud** [comment, marked as answer]
Hello

These messages say that the Jacobian is not fully implemented for the off-diagonal terms and that you may be missing some dependences in the diagonal term as well. 

You need to re-derive the Jacobian and implement the missing terms. 
Alternatively, you can switch your work to using automatic differentiation. This computes a perfect numerical Jacobian.
https://mooseframework.inl.gov/automatic_differentiation/#!

Guillaume

PS: there are also some caveats to the Jacobian analyzer. Make sure you followed the instructions for its use exactly otherwise it ll detect false positives

**Reply by Raven-pro** [reply]
Thanks！ I also want to know if there is any way to compute condition number of the matrix.
Is there any option of MOOSE to achieve that? :)

**Reply by GiudGiud** [reply]
Hello

yes you may use:
```
petsc_options_iname = '-pc_type'
petsc_options_ivalue = 'svd'
petsc_options = -pc_svd_monitor
```
this is really slow, use it only with a few thousands dofs at most

Guillaume

**Reply by Raven-pro** [reply]
Thank you!
